### PR TITLE
feat: Add strict mode to parser

### PR DESCRIPTION
**Pull request by aaronccasanova**
This PR introduces a `strict` mode parser config that is enabled by default.

### Errors on:
- Unknown option encountered
- Option of `type:'string'` used like a boolean option e.g. lone `--string`
- Option of `type:'boolean'` used like a string option e.g. `--boolean=foo`

### Examples:
```js
// node unknown-option.js --foo --bar
parseArgs({
  strict: true,
  options: { foo: { type: 'boolean' } },
});
// [Error]: Unknown option '--bar'
```

```js
// node invalid-string-option.js --foo
parseArgs({
  strict: true,
  options: { foo: { short: 'f', type: 'string' } },
})
// [Error]: Option '-f, --foo <value>' argument missing
```

```js
// node invalid-boolean-option.js --foo=bar
parseArgs({
  strict: true,
  options: { foo: { type: 'boolean' } },
})
// [Error]: Option '--foo' does not take an argument
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -58,8 +58,24 @@ function getMainArgs() {
...
-  const optionConfig = options[longOption] || {};
+function storeOptionValue(strict, options, longOption, value, result) {
+  const hasOptionConfig = ObjectHasOwn(options, longOption);
+
+  if (strict) {
+    if (!hasOptionConfig) {
+      throw new Error(`Unknown option: --${longOption}`);
+    }
+
+    if (options[longOption].type === 'string' && value == null) {
```
Comment (shadowspawn) [inline index.js:69]: I think it would be clearer to test against `undefined` rather than `null`, since the value is `undefined`.

Comment (aaronccasanova) [reply index.js]: https://github.com/pkgjs/parseargs/pull/74/commits/ca05b430e41c3056c367e6169329b5ea42d21d0b

Comment (ljharb) [reply index.js]: `== null` checks for both, which is the correct semantic check imo - just like `??` and optional chaining work.

Comment (shadowspawn) [reply index.js]: I understand that it covers both, but I still found it odd to test against `null` when that is not a value that we expect whereas `undefined` is. Good comparison with `??` and optional chaining. I didn't say how strongly to compare (and hadn't though that through!), and with that in mind suggest:

```
value == undefined
```

File: index.js [hunk 2]
```diff
@@ -66,11 +66,11 @@ function storeOptionValue(strict, options, longOption, value, result) {
       throw new Error(`Unknown option: --${longOption}`);
     }
 
-    if (options[longOption].type === 'string' && value == null) {
+    if (options[longOption].type === 'string' && value === undefined) {
```
[the remark this answers is not in the document]
Comment (shadowspawn) [reply index.js:69]: Either semantics work for me:
- `undefined` indicates failure to use correctly
- undefined indicates failure to use correctly

However, this leads me to wonder if a higher level test would be:
```
if (options[longOption].type === 'string' && typeof value !== 'string') {
```

How does that feel?

(I am not trying to imply that the "type" must be a core Javascript type, but it is in this case.)

File: index.js (hunk 1 above)
Comment (shadowspawn) [reply index.js]: Thanks, I had forgotten that possibility. (I haven't hit it in much code so not idiomatic in my mind yet!)

File: index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (shadowspawn) [reply index.js:69]: `value !== 'string'` would keep working as intended if we start passing in `true` in some cases too.

Also, I have come around to `value == null` being robust and idiomatic, and fine with that too with current code.

File: index.js (hunk 1 above)
Comment (shadowspawn) [reply index.js]: > since undefined is an identifier.

Checking my understanding. Perhaps not a practical problem in modern contexts since ECMAScript 5 made undefined read-only? (I was hacking about in node REPL and saw the read-only behaviour, went reading to find out why.)

Comment (shadowspawn) [reply index.js]: (I used `undefined` in a few other places in the code already outside this PR, which I will update to follow whatever pattern we use here.)

File: index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ljharb) [reply index.js:69]: that is a *much* better check, i think - programmatic users could pass any type of value, and we should be very conservative in what we expect.

File: index.js (hunk 1 above)
Comment (ljharb) [reply index.js]: ES5 made `undefined` readonly _only in the global scope_. it's fully overrideable everywhere else, including inside a CJS module's scope. The module wrapper can also be altered at runtime via `require.extensions`, so it's a real risk in node.

Comment (shadowspawn) [reply index.js]: (Thanks for the explanation. Appreciated.)

Comment (shadowspawn) [reply index.js]: (I will review use of `undefined` identifier outside `storeOptionValue` in #75, won't touch `storeOptionValue`.)

File: index.js [hunk 3]
```diff
@@ -58,8 +58,24 @@ function getMainArgs() {
...
+    if (!hasOptionConfig) {
+      throw new Error(`Unknown option: --${longOption}`);
+    }
+
+    if (options[longOption].type === 'string' && value == null) {
+      throw new Error(`Missing value for 'string' option: --${longOption}`);
+    }
+
+    if (options[longOption].type === 'boolean' && value != null) {
+      throw new Error(`Unexpected value for 'boolean' option: --${longOption}`);
```
[the remark this answers is not in the document]
Comment (shadowspawn) [reply index.js]: Light comment. I think we probably should have something to allow library users to catch and differentiate errors. I have not looked at this carefully, but see we have added `ERR_INVALID_ARG_TYPE`.

**Comment by bcoe** [comment]
@aaronccasanova bother you for a rebase?

File: test/index.js [hunk 4]
```diff
@@ -381,6 +381,68 @@ test('invalid union value passed to "type" option', function(t) {
   t.end();
 });
 
+// Test strict mode
+
+test('unknown long option --bar', function(t) {
+  const passedArgs = ['--foo', '--bar'];
```
Comment (bcoe) [inline test/index.js:387]: Should we default `strict` to `true`, like @ljharb has advocated?

@shadowspawn, @aaronccasanova what are your votes?

Comment (ljharb) [reply test/index.js:387]: Also @bakkot, i believe.

Comment (shadowspawn) [reply test/index.js:387]: https://github.com/pkgjs/parseargs/issues/11#issuecomment-1082664767

Comment (aaronccasanova) [reply test/index.js:387]: Yes, I vote `strict:true` by default. I like the idea of having `parseArgs` handle a majority cases off the shelf and allowing advance users/and library authors to explicitly opt out in favor of their own validation.

File: index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaronccasanova) [reply index.js]: Reverted back to [`==` and `!=`](https://github.com/pkgjs/parseargs/pull/74/commits/fa0775bb3f99a3eb52fbd60a97da028205d6d7f1)

File: test/index.js (hunk 4 above)
Comment (bcoe) [reply test/index.js:387]: The one ergonomic thing I don't like about strict by default, is that you end up having empty configuration:

```js
parseArgs({
  strict: true,
  options: {
    foo: {},
  },
});
```

If you decide you want the default `boolean` behavior.

What if we just don't make`type` optional. I prefer this because you also don't need to read docs to remember if `type` defaults to `string` or `boolean` (which I honestly forget).

Comment (aaronccasanova) [reply test/index.js:387]: > What if we just don't make `type` optional. I prefer this because you also don't need to read docs to remember if `type` defaults to `string` or `boolean` (which I honestly forget).

That's sensible to me. Similar to the behavior being `strict` for users the configuration would be `strict` for authors.

Comment (shadowspawn) [reply test/index.js:387]: I agree making type required is sensible from a readability point of view. Outside scope of this PR though?

File: index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (aaronccasanova) [reply index.js]: Added two new ERR_* classes and improved the error messages: https://github.com/pkgjs/parseargs/pull/74/commits/1d412dafd805ac9f334be666e4db18742411b06d

File: index.js (hunk 1 above)
Comment (aaronccasanova) [reply index.js:87]: https://github.com/pkgjs/parseargs/pull/74#discussion_r846578532

File: index.js [hunk 5]
```diff
@@ -73,8 +75,25 @@ function getMainArgs() {
...
+function storeOptionValue(strict, options, longOption, value, result) {
+  const hasOptionConfig = ObjectHasOwn(options, longOption);
+
+  if (strict) {
+    if (!hasOptionConfig) {
+      throw new ERR_UNKNOWN_OPTION(longOption);
+    }
+
+    if (options[longOption].type === 'string' && value == null) {
+      throw new ERR_INVALID_OPTION_VALUE(`Missing value for '${longOption}' option with type:'string'`);
```
Comment (shadowspawn) [inline index.js:88]: I think will have to pass in extra parameter to include the short option the user actually used in the error message. 
(The passed context has been steadily growing! I have been wondering about making the parameters an Object so the parameters are named in the calling code, or making the whole routine an internal function to `parseArgs` so has access to locals although that smacks a bit of globals and side-affects.)

File: errors.js [hunk 6]
```diff
@@ -23,7 +23,7 @@ class ERR_INVALID_SHORT_OPTION extends TypeError {
 
 class ERR_UNKNOWN_OPTION extends Error {
   constructor(longOption) {
-    super(`Unknown option: --${longOption}`);
+    super(`Unknown option '${longOption}'`);
```
Comment (shadowspawn) [inline errors.js:26]: This is where I tried `dashedOption` in another PR. 😄

Comment (aaronccasanova) [reply errors.js:26]: Ah yes, I like that! Would you prefer `dashedOption` or a more explicit `longOrShortOption` param?

File: index.js (hunk 5 above)
Comment (aaronccasanova) [reply index.js:88]: Agreed, it's starting to get unmanageable. We can probably clean things up with a `storeOptionCreator` higher-order function that holds the static values and returns a `storeOption` function for setting the dynamic bits.

```js
function parseArgs({
  strict = true,
  options = {},
}) {
 const result = {...}

  const storeOption = storeOptionCreator({
    strict,
    options,
    result,
  });

  if (isLongOption(arg)) {
    storeOption({ longOption, optionValue });
  }

  if (isShortOption(arg)) {
    storeOption({ shortOption, optionValue });
  }
}
```

Or simply make the parameters an Object as you suggested 👍

File: errors.js (hunk 6 above)
Comment (ljharb) [reply errors.js:26]: all options are dashed. long options are double-dashed.

File: index.js (hunk 5 above)
Comment (aaronccasanova) [reply index.js]: Went with an options bag for now and updated the error messages: https://github.com/pkgjs/parseargs/pull/74/commits/c4271a065b722e027601c94e019dc2e8a7a04434

File: errors.js [hunk 7]
```diff
@@ -7,16 +7,32 @@ class ERR_INVALID_ARG_TYPE extends TypeError {
   }
 }
 
+class ERR_INVALID_OPTION_VALUE extends Error {
```
Comment (bcoe) [inline errors.js:10]: As @[cjihrig](https://github.com/cjihrig) suggests elsewhere, could se use:

```
> new errors.codes.ERR_INVALID_ARG_VALUE('options.foo.short', '--bar', 'must be a single character');
```

To avoid needing to introduce a new error to `errors.js`?

File: errors.js [hunk 8]
```diff
@@ -7,16 +7,32 @@ class ERR_INVALID_ARG_TYPE extends TypeError {
...
+}
+
 class ERR_INVALID_SHORT_OPTION extends TypeError {
   constructor(longOption, shortOption) {
     super(`options.${longOption}.short must be a single character, got '${shortOption}'`);
     this.code = 'ERR_INVALID_SHORT_OPTION';
   }
 }
 
+class ERR_UNKNOWN_OPTION extends Error {
```
Comment (bcoe) [inline errors.js:24]: Likewise, if we can avoid adding any new errors it simplifies merging into Node.js.

File: index.js [hunk 9]
```diff
@@ -89,22 +119,24 @@ function storeOptionValue(options, longOption, value, result) {
...
       result.values[longOption] = [newValue];
   } else {
-    result.values[longOption] = value;
+    result.values[longOption] = optionValue;
   }
 }
 
 const parseArgs = ({
   args = getMainArgs(),
+  strict = false,
```
Comment (bcoe) [inline index.js:135]: Let's default this to `true` and update tests accordingly.

**Review by bcoe — CHANGES_REQUESTED** [review]
Left a couple nits, looking good to me.

File: errors.js (hunk 8 above)
Comment (bcoe) [reply errors.js:24]: see: https://github.com/pkgjs/parseargs/pull/97

File: errors.js (hunk 7 above)
Comment (aaronccasanova) [reply errors.js]: Here's an early WIP with all the strict error cases updated to use `ERR_INVALID_ARG_VALUE`: https://github.com/pkgjs/parseargs/pull/74/commits/948ca9e7270664cb9c558f040db5a29c743964e1

I could use some suggestions for the error messages as the new message constraint of `ERR_INVALID_ARG_VALUE` and my current implementation doesn't read entirely accurate:
```js
// node invalid-string-option.js --foo
parseArgs({
  strict: true,
  foo: { type: 'string' }
})
// TypeError: The property --foo must be provided a value, Received undefined.
```
```js
// node invalid-boolean-option.js --foo=bar
parseArgs({
  strict: true,
  foo: { type: 'boolean' }
})
// TypeError: The property --foo does not expect a value, Received bar.
```

Really I'm hung up on the word `property` over `option` (e.g. `The option --foo must...`), but not sure if I'm overthinking it. Just want the message to be super clear to users.

File: errors.js (hunk 8 above)
Comment (aaronccasanova) [reply errors.js]: https://github.com/pkgjs/parseargs/pull/74#discussion_r846813662

File: index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (aaronccasanova) [reply index.js]: Let's continue this conversation here: https://github.com/pkgjs/parseargs/pull/74#discussion_r846813662

File: errors.js (hunk 7 above)
Comment (aaronccasanova) [reply errors.js]: Similarly, the current [unknown option message](https://github.com/pkgjs/parseargs/blob/a2ea48de7ba3a5e48cfda50262cdf783bc672701/index.js#L92-L94) isn't entirely helpful to users e.g.
```sh
$ node unknown-option.js --foo

TypeError: The property option.foo must be configured, Received 'undefine'
```

Any suggestions to make this message more user-facing? e.g.
```
TypeError: The property --foo must be a known option, Received '--foo'
```

Comment (bakkot) [reply errors.js]: > Any suggestions to make this message more user-facing? 

As a user I would expect the message to be "Unknown option --foo".

Comment (aaronccasanova) [reply errors.js]: Agreed, that was the original message before the [rework](https://github.com/pkgjs/parseargs/commit/948ca9e7270664cb9c558f040db5a29c743964e1) with the addition of dynamism for long and short options e.g. `Unknown long option --foo` and `Unknown short option -f`. In an effort to reduce friction merging into Node I'm trying to find an equally helpful message within the constraints of `ERR_INVALID_ARG_VALUE`.

File: errors.js (hunk 8 above)
Comment (aaronccasanova) [reply errors.js]: Is it possible to namespace our error codes so we can provide the most accurate messages to users?
e.g.
- `ERR_PARGS_*`
- `ERR_PARGS_UNKNOWN_OPTION`
    - Unknown long option --foo
    - Unknown short option -f
-  `ERR_PARGS_INVALID_OPTION_VALUE` -
    - Missing value for ${longOrShortOption} with type:'string'
    - Unexpected value '${optionValue}' for ${longOrShortOption} with type:'boolean'

File: errors.js (hunk 7 above)
Comment (shadowspawn) [reply errors.js]: A quick comment: end-users do not know about the properties. They can see what they typed. So `--foo` and `-f` are  much more relevant than `foo`. And I think "option" is fine, do not need to specify "short" and "long", that is self evident.

(Disclaimer: I have not had a look yet at the constraints of `ERR_INVALID_ARG_VALUE` yet! Hopefully can go deeper tonight.)

Comment (shadowspawn) [reply errors.js]: I don't think the format for `ERR_INVALID_ARG_VALUE` is a particularly good match for our use here where we are referring to options on the command line, rather than arguments to a javascript function. (Valiant try though!)

I see the node 18.x branch has updated code for the error which inserts property/argument depending on the name, but still not a good match for our option errors.

https://github.com/nodejs/node/blob/ec5a359ffd38b6f56a456a0b6a482b5167fbfce1/lib/internal/errors.js#L1246

**Comment by bcoe** [comment]
@aaronccasanova landed @shadowspawn's refactor this morning, so I think this and @bakkot's potential refactor around positionals are the last two blockers for MVP.

File: errors.js (hunk 6 above)
Comment (shadowspawn) [reply errors.js]: Prefer `dashedOption` over `longOrShortOption`. However, I'll add another comment that perhaps not needed! (Which is a bit late given I suggest passing it in...)

**Comment by shadowspawn** [comment]
In Commander, the error messages refer to the option with both the short and long form and the argument if applicable, rather than just what was used on the command line. So in same style as the usage might appear in the help. If the user used the short form, the long form is more informative about the option purpose. 

```
% fab clone --branch
error: option '-b, --branch <branchname>' argument missing
% fab --silly
error: unknown option '--silly'
```

File: errors.js (hunk 6 above)
Comment (ljharb) [reply errors.js]: why not just `option`? i'm saying it's not an option unless it has a dash, so `dashedOption` is redundant.

File: errors.js (hunk 8 above)
Comment (aaronccasanova) [reply errors.js]: There appear to be error namespaces in node core:
- [ERR_IPC_*](https://github.com/nodejs/node/blob/ec5a359ffd38b6f56a456a0b6a482b5167fbfce1/lib/internal/errors.js#L1356-L1359)
- [ERR_MANIFEST_*](https://github.com/nodejs/node/blob/ec5a359ffd38b6f56a456a0b6a482b5167fbfce1/lib/internal/errors.js#L1360-L1391)
- [ERR_SOCKET_*](https://github.com/nodejs/node/blob/ec5a359ffd38b6f56a456a0b6a482b5167fbfce1/lib/internal/errors.js#L1512-L1529)
- [ERR_STREAM_*](https://github.com/nodejs/node/blob/ec5a359ffd38b6f56a456a0b6a482b5167fbfce1/lib/internal/errors.js#L1533-L1544)
- etc.

@bcoe @shadowspawn @shadowspawn What are your thoughts on using a namespace for `parseArgs` errors. For example `ERR_PARGS_*` or `ERR_PARSEARGS_*`

File: errors.js (hunk 7 above)
Comment (aaronccasanova) [reply errors.js:17]: Reverted to custom error classes and updated the error messages per: https://github.com/pkgjs/parseargs/pull/74#issuecomment-1095569158

Examples based on the PR description prompts:
- Unknown option encountered
```
Error: Unknown option '--foo'
Error: Unknown option '-f'
```
- Option of `type:'string'` used like a boolean option e.g. lone `--string`
```
Error: Option '--foo <foo>' argument missing
Error: `Option '-f, --foo <foo>' argument missing
```
- Option of `type:'boolean'` used like a string option e.g. `--boolean=foo`
```
Error: Option '--foo' does not take an argument
Error: `Option '-f, --foo' does not take an argument
```

File: index.js (hunk 9 above)
Comment (aaronccasanova) [reply index.js]: Started with a [naive approach](https://github.com/pkgjs/parseargs/pull/74/commits/e66a8b35714b693a200fe5dc2b3fde9541c3ae42) to get an understanding of how this change affects our existing tests:
- Defaulted `parseArgs` to `strict:true`
- Updated only failing tests to `strict:false`

I'm committing early for visibility but also I will likely need assistance updating/validating these tests. At first glance of the diff I'm wondering if we can clear things up by organizing these tests in a `strict-false.js` file or grouping `strict:false` tests in each file. I'm also fine with scoping that out to another PR.

cc @shadowspawn

File: test/index.js (hunk 4 above)
Comment (aaronccasanova) [reply test/index.js:407]: Feel free to reopen this conversation, but closing now as we all seem to be in agreement about `strict:true` default and explicitly setting option config types.

File: index.js (hunk 5 above)
Comment (aaronccasanova) [reply index.js]: Note: The [error messages have been updated again](https://github.com/pkgjs/parseargs/pull/74#discussion_r847846774) based on your feedback in https://github.com/pkgjs/parseargs/pull/74#issuecomment-1095569158

File: errors.js (hunk 6 above)
Comment (aaronccasanova) [reply errors.js]: This ended up only being referenced in one place so I inlined the conditional in the error: https://github.com/pkgjs/parseargs/blob/a4fc92af7cea085a016af01e812b812af87c7db9/index.js#L93

File: index.js [hunk 10]
```diff
@@ -73,16 +75,42 @@ function getMainArgs() {
...
+  const hasOptionConfig = ObjectHasOwn(options, longOption);
+
+  const optionConfig = hasOptionConfig ? options[longOption] : {};
+
+  if (strict) {
+    if (!hasOptionConfig) {
+      throw new ERR_UNKNOWN_OPTION(shortOption == null ? `--${longOption}` : `-${shortOption}`);
+    }
+
+    const shortOptionErr = optionConfig.short ? `-${optionConfig.short}, ` : '';
```
[the remark this answers is not in the document]
Comment (shadowspawn) [reply index.js:96]: It was validated as a single character on entry, so no need for extra checks here.

File: index.js (hunk 5 above)
Comment (shadowspawn) [reply index.js]: I tried the hybrid approach I had in mind (https://github.com/pkgjs/parseargs/pull/74#discussion_r846700566) and made an optional PR.

File: README.md [hunk 11]
```diff
@@ -97,24 +97,14 @@ process.mainArgs = process.argv.slice(process._exec ? 1 : 2)
...
-const args = ['-f', '--foo=a', '--bar', 'b'];
-const options = {};
-const { values, positionals } = parseArgs({ args, options });
-// values = { f: true, foo: 'a', bar: true }
-// positionals = ['b']
-```
-
 ```js
 const { parseArgs } = require('@pkgjs/parseargs');
 // type:string
```
Comment (shadowspawn) [inline README.md:102]:
```suggestion
// type:string and type:boolean
```

File: README.md [hunk 12]
```diff
@@ -97,24 +97,14 @@ process.mainArgs = process.argv.slice(process._exec ? 1 : 2)
...
 const args = ['-f', '--foo=a', '--bar', 'b'];
```
Comment (shadowspawn) [inline README.md:103]:
```suggestion
const args = ['--foo=a', '--bar', 'b'];
```

File: README.md (hunk 11 above)
Comment (shadowspawn) [reply README.md:102]: This example is a bit clumsy now with `f` declared as a slightly funky long/short to cover the previous implicit short. Making a series of suggestions to refactor a little.

Comment (shadowspawn) [reply README.md:102]: Going to be too messy, I'll make a single comment with code in an ordinary comment!

File: README.md [hunk 13]
```diff
@@ -97,24 +97,14 @@ process.mainArgs = process.argv.slice(process._exec ? 1 : 2)
...
 const options = {
-  bar: {
-    type: 'string',
-  },
+  f: { type: 'boolean' },
+  foo: { type: 'string'},
+  bar: { type: 'string' },
 };
 const { values, positionals } = parseArgs({ args, options });
 // values = { f: true, foo: 'a', bar: 'b' }
```
Comment (shadowspawn) [inline README.md:110]: `f` worked as an implicit boolean, but bit wild as a short-long. Suggest:

```
// specify the options that may be used
 const options = {
   foo: { type: 'string'},
   bar: { type: 'boolean' },
 };
 const args = ['--foo=a', '--bar'];
 const { values, positionals } = parseArgs({ args, options });
 // values = { foo: 'a', bar: true }
```

File: README.md [hunk 14]
```diff
@@ -126,6 +116,7 @@ const { parseArgs } = require('@pkgjs/parseargs');
 // type:string & multiple
 const args = ['-f', '--foo=a', '--foo', 'b'];
 const options = {
+  f: { type: 'boolean' },
   foo: {
     type: 'string',
     multiple: true,
```
Comment (shadowspawn) [inline README.md:122]: Same here, remove the funky `f` and trim down to the example of interest.

```
// type:string & multiple
const options = {
  foo: {
    type: 'string',
    multiple: true,
  },
};
const args = ['--foo=a', '--foo', 'b'];
const { values, positionals } = parseArgs({ args, options });
// values = { foo: [ 'a', 'b' ] }
// positionals = []
```

File: test/index.js [hunk 15]
```diff
@@ -401,6 +401,74 @@ test('invalid union value passed to "type" option', (t) => {
...
+  const passedOptions = { foo: { type: 'boolean' } };
+
+  t.throws(() => { parseArgs({ args: passedArgs, options: passedOptions }); }, {
+    code: 'ERR_UNKNOWN_OPTION'
+  });
+
+  t.end();
+});
+
+test('unknown option -r in short option group -bar', (t) => {
```
Comment (shadowspawn) [inline test/index.js:428]: Nice test

File: test/index.js [hunk 16]
```diff
@@ -401,6 +401,74 @@ test('invalid union value passed to "type" option', (t) => {
...
+  const passedOptions = { b: { type: 'boolean' }, a: { type: 'boolean' } };
+
+  t.throws(() => { parseArgs({ args: passedArgs, options: passedOptions }); }, {
+    code: 'ERR_UNKNOWN_OPTION'
+  });
+
+  t.end();
+});
+
+test('unknown option with explicit value', (t) => {
```
Comment (shadowspawn) [inline test/index.js:439]: nice test

File: index.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (aaronccasanova) [reply index.js:96]: The initial validation only checks own properties of the option config. Couldn't immediatly think of a way to confirm, but came up with this demo:
image

Not sure if this opens up any bugs or areas for malicious behavior, but might as well update the condition to:
```js
const shortOptionErr = ObjectHasOwn(optionConfig, 'short') ? `-${optionConfig.short}, ` : '';
```

File: errors.js (hunk 7 above)
Comment (aaronccasanova) [reply errors.js:17]: Good point! That example is excessively long. I think I'll use `value` for consistency with the `parseArgs` results object.

File: index.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (aaronccasanova) [reply index.js]: https://github.com/pkgjs/parseargs/pull/74/commits/12fd0e6d58fda45d49f2d310cf4ec0090e03fcbe

File: errors.js (hunk 7 above)
Comment (aaronccasanova) [reply errors.js]: https://github.com/pkgjs/parseargs/pull/74/commits/870cffefd7d1cae28564c374b23eb21e6fb6c3c1

File: errors.js (hunk 8 above)
Comment (aaronccasanova) [reply errors.js]: Updated unique `parseArgs` error classes to `ERR_PARSE_ARGS_*`: https://github.com/pkgjs/parseargs/pull/74/commits/694c75d849e60f470d7c5aaa0fe929f9e64a7249

File: README.md (hunk 13 above)
Comment (aaronccasanova) [reply README.md]: * https://github.com/pkgjs/parseargs/pull/74/commits/9b76d55779372e20cddbc67284787a4283f9a14f
* https://github.com/pkgjs/parseargs/pull/74/commits/e56246c59179b3e1ca38654cde68636c05ac6c0e

File: README.md (hunk 14 above)
Comment (aaronccasanova) [reply README.md]: * https://github.com/pkgjs/parseargs/pull/74/commits/9b76d55779372e20cddbc67284787a4283f9a14f
* https://github.com/pkgjs/parseargs/pull/74/commits/e56246c59179b3e1ca38654cde68636c05ac6c0e

File: index.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (shadowspawn) [reply index.js]: Arg, you are right! (Sorry, the validation wasn't doing what I thought.) 

I assume the high level goal to behave the same as if the prototype pollution was not present?

That makes me question lots more of the code though. The options bag has introduced lots of property access:
- same issue for `multiple`
- and `type`, lots including utils. Although that might be ok if the next issue was fixed.
- the validation of `type` at the top of `parseArgs` does not test it is not from prototype (from my recent change to make type required!)
- the destructing function parameters of both `parseArgs` and (now) `storeOption` receive properties from prototype if not specified by caller.

We could copy the input arguments into a safe(r) object before using them, by specifying all expected properties. Is this a pattern used in other node code? I suspect the usual pattern is just be paranoid everywhere because easier to be confident local code is being paranoid enough?!

I suggest only fix up the code introduced in this PR though. Get `strict` functional and separately have an more-paranoia-more-robust revisit!

File: index.js [hunk 17]
```diff
@@ -73,16 +75,42 @@ function getMainArgs() {
...
+  strict,
+  options,
+  result,
+  longOption,
+  shortOption,
+  optionValue,
+}) {
+  const hasOptionConfig = ObjectHasOwn(options, longOption);
+
+  const optionConfig = hasOptionConfig ? options[longOption] : {};
```
Comment (bcoe) [inline index.js:89]: I might simplify this to `options[longOption] ?? {}`.

**Review by bcoe — CHANGES_REQUESTED** [review]
Looking good to me, mostly left nits.


Only blocking from me, I prefer to check explicitly for `=== undefined`, `!== undefined`, rather than a loose comparison.

Thanks for doing this work.

File: index.js [hunk 18]
```diff
@@ -73,16 +75,42 @@ function getMainArgs() {
...
+
+  if (strict) {
+    if (!hasOptionConfig) {
+      throw new ERR_PARSE_ARGS_UNKNOWN_OPTION(shortOption == null ? `--${longOption}` : `-${shortOption}`);
```
[the remark this answers is not in the document]
Comment (shadowspawn) [reply index.js:93]: Me too! But we had some _long_ conversions about this, and the reasoning for doing weak test against null to avoid contamination of undefined.

- https://github.com/pkgjs/parseargs/pull/74#discussion_r818969649

[the remark this answers is not in the document]
Comment (shadowspawn) [reply index.js:93]: Also: https://github.com/pkgjs/parseargs/pull/74#discussion_r819093456

[the remark this answers is not in the document]
Comment (ljharb) [reply index.js:93]: The point is to check for both; that’s why the comparison is loose.

[the remark this answers is not in the document]
Comment (bcoe) [reply index.js:93]: works for me.

**Review by bcoe — APPROVED** [review]

File: index.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (shadowspawn) [reply index.js]: @bcoe @ljharb 
If my reading of this is correct, we are not currently robust against prototype pollution. Can we improve this after first version of a prototype feature?

(I have some ideas about how to improve it without littering the code and can have a go tonight, but clock is ticking.)

**Review by shadowspawn — APPROVED** [review]

**Comment by aaronccasanova** [comment]
Thanks for the reviews and feedback! I'm happy for this PR to be merged as is and willing to help with any quick follow ups. My last PR was merged by @shadowspawn and I'm not entirely sure if I should be merging into main. That being said, I don't want to block anyone, so feel free to merge and/or push up any updates!

**Comment by shadowspawn** [comment]
Taking a last look now. Thanks for all the hard work @aaronccasanova

File: index.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (bcoe) [reply index.js]: I think the bigger concern for prototype pollution is untrusted input to the CLI modifying the prototype, vs., the user having mucked with __proto__ in their application, and this affecting our parse.

This makes me think that the bigger security concern would be if we ever supported dot properties; as @shadowspawn has mentioned.
